### ISSUE TITLE
Update recurrent weights for spiking neural networks

### ISSUE THREAD

**Issue by gogoARS** [zhenruiliao/tension#36]
Thanks for such a nice library of FORCE learning! This is very helpful!

I think currently tension only supports the learning of readout weights for spiking neural networks. I was wondering if there's a way to also update the recurrent weights during the training of SNN, similar to what full-FORCE did? And where can I customize the learning steps if needed?

**Comment by lubin-liu** [comment]
By design the user is not allowed to update the recurrent weight for SNNs. This is because the SNN just an implementation of  [this paper](https://www.nature.com/articles/s41467-017-01827-3). If you go to equation 13, the recurrent weight is broken down into a static component and a learned decoder (that is a function of the readout weight), so the recurrent weight is updated when the readout weight is updated. For SNNs in this package, the 'recurrent kernel' is just the static component of the weight matrix times G, which isn't supposed to be updated, as the actual recurrent weight is updated indirectly when the readout weight is updated.

Could you clarify what you mean by learning step?

**Comment by gogoARS** [comment]
Thank you very much for the quick reply! 

> By design the user is not allowed to update the recurrent weight for SNNs. This is because the SNN just an implementation of [this paper](https://www.nature.com/articles/s41467-017-01827-3). If you go to equation 13, the recurrent weight is broken down into a static component and a learned decoder (that is a function of the readout weight), so the recurrent weight is updated when the readout weight is updated. For SNNs in this package, the 'recurrent kernel' is just the static component of the weight matrix times G, which isn't supposed to be updated, as the actual recurrent weight is updated indirectly when the readout weight is updated.

For SNN, I wonder if this indirectly updated recurrent weights are used to compute the neuron voltage in the next timestep in tension? I checked the original implementation of the nature communications paper, but only found that the static recurrent weight itself is used to compute the postsynaptic current at every timestep.

> Could you clarify what you mean by learning step?

Sorry for the possible misunderstand here. By learning I mean the training process, like the train_step function of the FullFORCEmodel. Is it possible to define a customized SNNModel and write my own train_step to introduce the update of recurrent weight for spiking neural nerworks, separately from the readout weight?

**Comment by lubin-liu** [comment]
> For SNN, I wonder if this indirectly updated recurrent weights are used to compute the neuron voltage in the next timestep in tension? I checked the original implementation of the nature communications paper, but only found that the static recurrent weight itself is used to compute the postsynaptic current at every timestep.

From Eq. 11 of the Nature paper, the synaptic current is updated using the recurrent weight, not just the static recurrent weight, which is how it's done with `tension` (see [`compute_current` method](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L185), noting that `self.recurrent_kernel` is the static recurrent weight * G).  

> Sorry for the possible misunderstand here. By learning I mean the training process, like the train_step function of the FullFORCEmodel. Is it possible to define a customized SNNModel and write my own train_step to introduce the update of recurrent weight for spiking neural nerworks, separately from the readout weight?

Note that in the implementation here there's no explicit variable defining the recurrent weight from the paper; you would have to construct it based on Eq. 13 using the other variables.  If you wanted to update the static recurrent weight, first you would have to hack [this line](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L84) by changing `recurrent_kernel_trainable=True` (I hard coded this since the static weight is not supposed to be trainable). You can then sub-class `SpikingNNModel` and tweak the `train_step` based [here](https://zhenruiliao.github.io/tension/usage.html#customizing-train-step). In your case it seems like you would want to write your own custom `update_recurrent_kernel` so the highlighted part would have to be updated

![Capture](https://github.com/zhenruiliao/tension/assets/69323872/3000fc40-2f96-45cb-9dea-5cf18fcaab8b)

**Comment by gogoARS** [comment]
> From Eq. 11 of the Nature paper, the synaptic current is updated using the recurrent weight, not just the static recurrent weight, which is how it's done with tension (see [compute_current method](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L185), noting that self.recurrent_kernel is the static recurrent weight * G).

Thanks for the explanations! 

One more question: In [`compute_current`](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L185), the current is computed as `self.I_bias + backend.dot(h, self.recurrent_kernel) + backend.dot(out, self.feedback_kernel) + backend.dot(inputs, self.input_kernel)`. I can see the bias term, static recurrent weight related term, feedback term and input term here. Since the trained readout weight `self.output_kernel` is also served as the learnt weight for the synaptic weight $\omega$, I wonder how `self.output_kernel` is used here? I cannot find it in [`compute_current`](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L185).

In addition, I found that `compute_current` is overriden in [`OptimizedSpikingNN`](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L263):
`self.I_bias + ipsc + backend.dot(out, self.feedback_kernel) + backend.dot(inputs, self.input_kernel)`. Here `ipsc` replaces `backend.dot(h, self.recurrent_kernel)` and is computed with `jd`, where `jd = tf.math.reduce_sum(self.recurrent_kernel[v_mask[0] == 1], axis=0, keepdims=True)`. I still cannot see how the learnt recurrent weight contributed by `self.output_kernel` is applied here. 

I'm confused with the computation here and I would very much appreciate it if you could explain more about it! Thank you!

**Comment by lubin-liu** [comment]
The output kernel is used in `backend.dot(out, self.feedback_kernel)`, where `out` is the output of the network or equivalently, the firing rate of each neuron times the read out weights as computed from [this line in the forward pass](https://github.com/zhenruiliao/tension/blob/17ae8e00132ed66b3cfc1f512deeacf2a41a9b2c/src/tension/spiking.py#L207) and retained as a SNN layer state. This is effectively the second part of Eq. 13 multiplied by the neuron firing rate; I re-used the previously computed network output to save on computation. The same applies to your second question regarding `OptimizedSpikingNN`.

**Comment by gogoARS** [comment]
Ah I see. So basically the learnt readout weight is indirectly used as the synaptic weight of the SNN by adding the feedback part. Thank you for the explanations. This is super helpful!
